server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceService.java

```diff
@@ -244,7 +244,8 @@ boolean updateDesiredBalanceAndReroute(DesiredBalanceInput desiredBalanceInput,
                 || ignored.unassignedInfo().getLastAllocationStatus() == UnassignedInfo.AllocationStatus.NO_ATTEMPT
                 : "Unexpected status: " + ignored.unassignedInfo().getLastAllocationStatus();
 
-            assignments.merge(ignored.shardId(), ShardAssignment.UNASSIGNED, ShardAssignment::merge);
+            var unassigned = ignored.unassignedInfo().getLastAllocationStatus() == UnassignedInfo.AllocationStatus.DECIDERS_NO;
+            assignments.merge(ignored.shardId(), unassigned ? ShardAssignment.UNASSIGNED : ShardAssignment.IGNORED, ShardAssignment::merge);
```

Comment:
It looks like when restoring shards from the snapshots they are ignored for some amount of time until their size could be estimated. We need to keep this information in desired balance to mark them as NO_ATTEMPT until this information is available.

Reply:
This should fix multiple (if not all) snapshot related integration tests.

Reply:
I am not sure I follow. `compute` seems to be more verbose as it always calls the `remappingFunction` and require null handling for previous value (when missing).  Could you please clarify why you would like to change this?